File: lib/signalgrid-simulator/src/decisionEngine.ts
import { createAuditEvidence } from "./audit";
import { routeDecision } from "./routing";
import { simulatorScenarios } from "./scenarios";
import type {
  DecisionOutcome,
  SignalGridDecision,
  SignalGridSignal,
  SimulatorRunResult,
  SimulatorScenario,
} from "./types";

const evaluationTime = "2026-06-09T14:05:00.000Z";

export function listSimulatorScenarios(): SimulatorScenario[] {
  return simulatorScenarios;
}

export function getSimulatorScenario(
  id: string,
): SimulatorScenario | undefined {
  return simulatorScenarios.find((scenario) => scenario.id === id);
}

export function runSimulatorScenario(scenarioId: string): SimulatorRunResult {
  const scenario = getSimulatorScenario(scenarioId);
  if (!scenario) {
    throw new Error(`Unknown simulator scenario: ${scenarioId}`);
  }

  return runScenario(scenario);
}

export function runScenario(scenario: SimulatorScenario): SimulatorRunResult {
  const normalizedSignals = normalizeSignals(scenario.startingSignals);
  const decision = evaluateScenario(scenario, normalizedSignals);
  const routedActions = routeDecision(scenario, decision.outcomes);
  const auditEvidence = createAuditEvidence(scenario, decision, routedActions);
  const status =
    scenario.expectedOutcomes.every((outcome) =>
      decision.outcomes.includes(outcome),
    ) && auditEvidence.length > 0
      ? "PASS"
      : "FAIL";

  return {
    scenario,
    normalizedSignals,
    decision,
    routedActions,
    auditEvidence,
    status,
  };
}

export function normalizeSignals(
  signals: SignalGridSignal[],
): SignalGridSignal[] {
  return signals.map((signal) => ({
    ...signal,
    attributes: {
      ...signal.attributes,
      normalized: true,
      simulatorOnly: true,
    },
  }));
}

function evaluateScenario(
  scenario: SimulatorScenario,
  signals: SignalGridSignal[],
): SignalGridDecision {
  const hasType = (type: SignalGridSignal["type"]) =>
    signals.some((signal) => signal.type === type);
  const hasCriticalOrHighSecurityRisk = signals.some(
    (signal) =>
      signal.type === "identity.risk_detected" ||
      signal.attributes["securityRisk"] === "high" ||
      signal.attributes["edr"] === "disabled" ||
      signal.attributes["edrRisk"] === "high",
  );
  const hasIdentityIntegrityFailure = signals.some(
    (signal) =>
      signal.layer === "identity" &&
      (signal.attributes["identityState"] === "invalid" ||
        signal.attributes["identityState"] === "disabled" ||
        signal.attributes["risk"] === "high"),
  );
  const hasDeviceTrustFailure = signals.some(
    (signal) =>
      signal.attributes["compliance"] === "non_compliant" ||
      signal.attributes["managementState"] === "unmanaged" ||
      signal.attributes["deviceCompromised"] === true ||
      signal.attributes["sourceIntegrity"] === "failed",
  );
  const hasStateFreshnessFailure = signals.some(
    (signal) =>
      signal.attributes["declaredState"] === "stale" ||
      signal.attributes["ssoStatus"] === "missing" ||
      signal.attributes["freshness"] === "stale" ||
      signal.attributes["criticalInput"] === "malformed",
  );
  const hasCustodyFailure = signals.some(
    (signal) =>
      signal.attributes["zone"] === "wrong" ||
      signal.attributes["location"] === "unknown" ||
      signal.attributes["missingReturn"] === true ||
      signal.attributes["overdue"] === true ||
      signal.attributes["returnBay"] === "wrong",
  );
  const hasWorkflowRoutingFailure = signals.some(
    (signal) =>
      signal.attributes["workflowOwner"] === "missing" ||
      signal.attributes["requiredApproval"] === "missing" ||
      signal.attributes["escalationDestination"] === "unavailable",
  );
  const hasActiveCustodyIntegrityFailure =
    hasCustodyFailure ||
    hasType("dock.device_missing") ||
    hasType("dock.wrong_slot_return") ||
    hasType("rtls.wrong_zone") ||
    hasType("rts.staff_safety_alert");
  const outcomes = new Set<DecisionOutcome>();
  const reasonCodes: string[] = [];

  if (hasIdentityIntegrityFailure) {
    outcomes.add("restrict");
    outcomes.add("alert_operator");
    outcomes.add("route_to_owner");
    reasonCodes.push("IDENTITY_INTEGRITY_FAILURE");
  }

  if (hasDeviceTrustFailure) {
    outcomes.add("restrict");
    outcomes.add("create_ticket");
    outcomes.add("alert_operator");
    reasonCodes.push("DEVICE_TRUST_FAILURE");
  }

  if (hasStateFreshnessFailure) {
    outcomes.add("step_up");
    outcomes.add("request_remediation");
    reasonCodes.push("STATE_FRESHNESS_FAILURE");
  }

  if (hasCustodyFailure) {
    outcomes.add("create_ticket");
    outcomes.add("alert_operator");
    outcomes.add("route_to_owner");
    reasonCodes.push("CUSTODY_EXCEPTION");
  }

  if (hasWorkflowRoutingFailure) {
    outcomes.add("step_up");
    outcomes.add("alert_operator");
    outcomes.add("route_to_owner");
    reasonCodes.push("WORKFLOW_ROUTE_UNAVAILABLE");
  }

  if (hasType("device.non_compliant")) {
    outcomes.add("restrict");
    outcomes.add("create_ticket");
    outcomes.add("alert_operator");
    reasonCodes.push("DEVICE_NON_COMPLIANT");
  }

  if (hasType("device.stale_checkin")) {
    outcomes.add("step_up");
    outcomes.add("request_remediation");
    reasonCodes.push("POSTURE_STALE");
  }

  if (hasType("dock.device_missing") || hasType("dock.wrong_slot_return")) {
    outcomes.add("create_ticket");
    outcomes.add("alert_operator");
    outcomes.add("route_to_owner");
    reasonCodes.push("DOCK_EXCEPTION");
  }

  if (hasType("rtls.wrong_zone") || hasType("rts.staff_safety_alert")) {
    outcomes.add("alert_operator");
    outcomes.add("route_to_owner");
    reasonCodes.push("LOCATION_EXCEPTION");
  }

  if (hasType("device.low_battery")) {
    outcomes.add("route_to_owner");
    outcomes.add("alert_operator");
    reasonCodes.push("BATTERY_WORKFLOW_RISK");
  }

  if (hasType("device.health_degraded")) {
    outcomes.add("route_to_owner");
    outcomes.add("create_ticket");
    reasonCodes.push("OPERATIONAL_HEALTH_DEGRADED");
  }

  if (hasCriticalOrHighSecurityRisk) {
    outcomes.add("restrict");
    outcomes.add("alert_operator");
    outcomes.add("route_to_owner");
    reasonCodes.push("SECURITY_RISK_ESCALATION");
  }

  if (hasType("api.integration_failed")) {
    outcomes.add("alert_operator");
    outcomes.add("route_to_owner");
    reasonCodes.push("INTEGRATION_ROUTE_DEGRADED");
  }

  if (hasType("remediation.verified")) {
    outcomes.add("verify_remediation");
    outcomes.add("allow");
    reasonCodes.push("REMEDIATION_VERIFIED");
  }

  if (
    outcomes.size === 0 &&
    hasType("identity.authenticated") &&
    (hasType("device.posture_observed") ||
      hasType("apple.ddm_declared_state")) &&
    !hasType("device.non_compliant")
  ) {
    outcomes.add("allow");
    reasonCodes.push(
      hasType("apple.ddm_declared_state")
        ? "APPLE_DECLARED_STATE_TRUSTED"
        : "IDENTITY_AND_POSTURE_TRUSTED",
    );
  }

  outcomes.add("record_audit");

  if (
    outcomes.has("allow") &&
    (outcomes.has("restrict") ||
      outcomes.has("deny") ||
      outcomes.has("step_up"))
  ) {
    outcomes.delete("allow");
    reasonCodes.push("ALLOW_REMOVED_DUE_TO_HIGHER_RISK");
  }

  if (outcomes.has("allow") && hasActiveCustodyIntegrityFailure) {
    outcomes.delete("allow");
    reasonCodes.push("ALLOW_REMOVED_DUE_TO_CUSTODY_FAILURE");
  }

  const ordered = orderOutcomes([...outcomes]);

  return {
    id: `decision:${scenario.id}`,
    scenarioId: scenario.id,
    evaluatedAt: evaluationTime,
    outcomes: ordered,
    primaryOutcome: ordered[0] ?? "record_audit",
    confidence: reasonCodes.includes("INTEGRATION_ROUTE_DEGRADED")
      ? "medium"
      : "high",
    reasonCodes,
    explanation: explainDecision(ordered, reasonCodes),
  };
}

function orderOutcomes(outcomes: DecisionOutcome[]): DecisionOutcome[] {
  const order: DecisionOutcome[] = [
    "deny",
    "restrict",
    "step_up",
    "alert_operator",
    "create_ticket",
    "route_to_owner",
    "request_remediation",
    "verify_remediation",
    "allow",
    "record_audit",
  ];
  return order.filter((outcome) => outcomes.includes(outcome));
}

function explainDecision(
  outcomes: DecisionOutcome[],
  reasonCodes: string[],
): string {
  if (outcomes.includes("restrict")) {
    return "Risk exceeded the simulator trust threshold, so the workflow is restricted and routed for review.";
  }
  if (outcomes.includes("step_up")) {
    return "Trust is incomplete because posture freshness is stale, so the simulator requests review or refresh before treating the device as fully trusted.";
  }
  if (
    outcomes.includes("alert_operator") ||
    outcomes.includes("create_ticket")
  ) {
    return `Operational exception routed with reason codes: ${reasonCodes.join(", ")}.`;
  }
  if (outcomes.includes("verify_remediation")) {
    return "Remediation evidence is verified in the fixture, so the device becomes an allow candidate with audit evidence.";
  }
  return "Identity, posture, location, dock, and workflow signals are aligned in the fixture, so the simulator allows the candidate workflow and records audit evidence.";
}

File: native/ios/EnterpriseShell/Services/DecisionEngine.swift
import Foundation

/// Native port of the deterministic simulator decision engine
/// (`lib/signalgrid-simulator/src/decisionEngine.ts`, per `SIMULATOR_DECISION_ENGINE.md`).
///
/// Maps a set of normalized signals (identity, device posture/compliance, state
/// freshness, custody/location, dock, workflow, security risk) to a primary gate
/// outcome + reason codes. Deterministic and fail-closed. This replaces hand-set
/// per-action verdicts: the host app builds a signal context and the engine
/// decides, exactly as the sim suite does — so the on-device flow reflects the
/// real decision logic, not a script.
enum DecisionEngine {

    struct Signal {
        let type: String
        var layer: String? = nil
        var attributes: [String: String] = [:]

        init(_ type: String, layer: String? = nil, attributes: [String: String] = [:]) {
            self.type = type; self.layer = layer; self.attributes = attributes
        }

        // Convenience signal contexts (public-safe, synthetic).
        static let authenticated = Signal("identity.authenticated")
        static let postureObserved = Signal("device.posture_observed")
        static let staleCheckin = Signal("device.stale_checkin")   // → step_up (POSTURE_STALE)
        static let nonCompliant = Signal("device.non_compliant")   // → restrict (DEVICE_NON_COMPLIANT)
        static let wrongZone = Signal("rtls.wrong_zone")           // custody/location exception
        static let securityRisk = Signal("identity.risk_detected") // → restrict (SECURITY_RISK_ESCALATION)
    }

    struct Result {
        let outcome: AppWorkflows.DecisionOutcome   // gate outcome fed to the planner
        let reasonCodes: [String]
        let explanation: String
        let allOutcomes: [String]                   // full ordered simulator outcome set
    }

    static func evaluate(_ signals: [Signal]) -> Result {
        func hasType(_ t: String) -> Bool { signals.contains { $0.type == t } }

        let hasCriticalOrHighSecurityRisk = signals.contains {
            $0.type == "identity.risk_detected" ||
            $0.attributes["securityRisk"] == "high" ||
            $0.attributes["edr"] == "disabled" ||
            $0.attributes["edrRisk"] == "high"
        }
        let hasIdentityIntegrityFailure = signals.contains {
            $0.layer == "identity" &&
            ($0.attributes["identityState"] == "invalid" ||
             $0.attributes["identityState"] == "disabled" ||
             $0.attributes["risk"] == "high")
        }
        let hasDeviceTrustFailure = signals.contains {
            $0.attributes["compliance"] == "non_compliant" ||
            $0.attributes["managementState"] == "unmanaged" ||
            $0.attributes["deviceCompromised"] == "true" ||
            $0.attributes["sourceIntegrity"] == "failed"
        }
        let hasStateFreshnessFailure = signals.contains {
            $0.attributes["declaredState"] == "stale" ||
            $0.attributes["ssoStatus"] == "missing" ||
            $0.attributes["freshness"] == "stale" ||
            $0.attributes["criticalInput"] == "malformed"
        }
        let hasCustodyFailure = signals.contains {
            $0.attributes["zone"] == "wrong" ||
            $0.attributes["location"] == "unknown" ||
            $0.attributes["missingReturn"] == "true" ||
            $0.attributes["overdue"] == "true" ||
            $0.attributes["returnBay"] == "wrong"
        }
        let hasWorkflowRoutingFailure = signals.contains {
            $0.attributes["workflowOwner"] == "missing" ||
            $0.attributes["requiredApproval"] == "missing" ||
            $0.attributes["escalationDestination"] == "unavailable"
        }
        let hasActiveCustodyIntegrityFailure = hasCustodyFailure ||
            hasType("dock.device_missing") || hasType("dock.wrong_slot_return") ||
            hasType("rtls.wrong_zone") || hasType("rts.staff_safety_alert")

        var outcomes = Set<String>()
        var reasonCodes: [String] = []

        if hasIdentityIntegrityFailure {
            outcomes.formUnion(["restrict", "alert_operator", "route_to_owner"]); reasonCodes.append("IDENTITY_INTEGRITY_FAILURE")
        }
        if hasDeviceTrustFailure {
            outcomes.formUnion(["restrict", "create_ticket", "alert_operator"]); reasonCodes.append("DEVICE_TRUST_FAILURE")
        }
        if hasStateFreshnessFailure {
            outcomes.formUnion(["step_up", "request_remediation"]); reasonCodes.append("STATE_FRESHNESS_FAILURE")
        }
        if hasCustodyFailure {
            outcomes.formUnion(["create_ticket", "alert_operator", "route_to_owner"]); reasonCodes.append("CUSTODY_EXCEPTION")
        }
        if hasWorkflowRoutingFailure {
            outcomes.formUnion(["step_up", "alert_operator", "route_to_owner"]); reasonCodes.append("WORKFLOW_ROUTE_UNAVAILABLE")
        }
        if hasType("device.non_compliant") {
            outcomes.formUnion(["restrict", "create_ticket", "alert_operator"]); reasonCodes.append("DEVICE_NON_COMPLIANT")
        }
        if hasType("device.stale_checkin") {
            outcomes.formUnion(["step_up", "request_remediation"]); reasonCodes.append("POSTURE_STALE")
        }
        if hasType("dock.device_missing") || hasType("dock.wrong_slot_return") {
            outcomes.formUnion(["create_ticket", "alert_operator", "route_to_owner"]); reasonCodes.append("DOCK_EXCEPTION")
        }
        if hasType("rtls.wrong_zone") || hasType("rts.staff_safety_alert") {
            outcomes.formUnion(["alert_operator", "route_to_owner"]); reasonCodes.append("LOCATION_EXCEPTION")
        }
        if hasType("device.low_battery") {
            outcomes.formUnion(["route_to_owner", "alert_operator"]); reasonCodes.append("BATTERY_WORKFLOW_RISK")
        }
        if hasType("device.health_degraded") {
            outcomes.formUnion(["route_to_owner", "create_ticket"]); reasonCodes.append("OPERATIONAL_HEALTH_DEGRADED")
        }
        if hasCriticalOrHighSecurityRisk {
            outcomes.formUnion(["restrict", "alert_operator", "route_to_owner"]); reasonCodes.append("SECURITY_RISK_ESCALATION")
        }
        if hasType("api.integration_failed") {
            outcomes.formUnion(["alert_operator", "route_to_owner"]); reasonCodes.append("INTEGRATION_ROUTE_DEGRADED")
        }
        if hasType("remediation.verified") {
            outcomes.formUnion(["verify_remediation", "allow"]); reasonCodes.append("REMEDIATION_VERIFIED")
        }

        if outcomes.isEmpty && hasType("identity.authenticated") &&
            (hasType("device.posture_observed") || hasType("apple.ddm_declared_state")) &&
            !hasType("device.non_compliant") {
            outcomes.insert("allow")
            reasonCodes.append(hasType("apple.ddm_declared_state") ? "APPLE_DECLARED_STATE_TRUSTED" : "IDENTITY_AND_POSTURE_TRUSTED")
        }

        outcomes.insert("record_audit")

        if outcomes.contains("allow") &&
            (outcomes.contains("restrict") || outcomes.contains("deny") || outcomes.contains("step_up")) {
            outcomes.remove("allow"); reasonCodes.append("ALLOW_REMOVED_DUE_TO_HIGHER_RISK")
        }
        if outcomes.contains("allow") && hasActiveCustodyIntegrityFailure {
            outcomes.remove("allow"); reasonCodes.append("ALLOW_REMOVED_DUE_TO_CUSTODY_FAILURE")
        }

        let order = ["deny", "restrict", "step_up", "alert_operator", "create_ticket",
                     "route_to_owner", "request_remediation", "verify_remediation", "allow", "record_audit"]
        let ordered = order.filter { outcomes.contains($0) }

        return Result(outcome: gateOutcome(ordered),
                      reasonCodes: reasonCodes,
                      explanation: explain(ordered, reasonCodes),
                      allOutcomes: ordered)
    }

    /// Reduce the full simulator outcome set to the four gate outcomes the
    /// app-workflows planner consumes. Fail-closed: if only routing/audit outcomes
    /// remain (no allow), hold for a step-up (mirrors NO_RULE_MATCHED_DEFAULT_STEP_UP).
    private static func gateOutcome(_ ordered: [String]) -> AppWorkflows.DecisionOutcome {
        if ordered.contains("deny") { return .deny }
        if ordered.contains("restrict") { return .restrict }
        if ordered.contains("step_up") { return .step_up }
        if ordered.contains("allow") { return .allow }
        return .step_up
    }

    private static func explain(_ outcomes: [String], _ reasonCodes: [String]) -> String {
        if outcomes.contains("restrict") {
            return "Risk exceeded the trust threshold, so the workflow is restricted and routed for review."
        }
        if outcomes.contains("step_up") {
            return "Trust is incomplete (posture freshness), so review/refresh is required before treating the device as fully trusted."
        }
        if outcomes.contains("alert_operator") || outcomes.contains("create_ticket") {
            return "Operational exception routed: \(reasonCodes.joined(separator: ", "))."
        }
        if outcomes.contains("verify_remediation") {
            return "Remediation verified → allow candidate with audit evidence."
        }
        return "Identity, posture, location, dock, and workflow signals are aligned, so the workflow is allowed with audit evidence."
    }
}
